### ISSUE TITLE
Chained Tasks Not Executed In-order

### ISSUE THREAD

**Issue by arshad171** [celery/celery#7117]
### Discussed in https://github.com/celery/celery/discussions/7114

<div type='discussions-op-text'>

<sup>Originally posted by **arshad171** November 22, 2021</sup>
For some of my TensorFlow pipelines, I need the task chains to be executed **in the order they were received**, i.e. execute the entire chain before beginning with the next chain scheduled.

I have abstracted the problem with the example below,

## Sample Tasks

```
@app.task
def task1(id):
    print(f'{id=} | task 1')
    time.sleep(5)
    return id

@app.task
def task2(id):
    print(f'{id=} | task 2')
    time.sleep(5)
    return id
```

## Chaining Tasks

Creating two chains, with different ids, `id=1` and `id=2`.

```
res1 = chain(
    task1.s(1),
    task2.s()
).apply_async()

res2 = chain(
    task1.s(2),
    task2.s()
).apply_async()
```

### The result I expect: The first chain (with id=1) to be processed before the chain with id=2 is picked up.

```
[2021-11-22 20:42:43,653: WARNING/ForkPoolWorker-1] id=1 | task 1
[2021-11-22 20:42:48,663: WARNING/ForkPoolWorker-1] id=1 | task 2
[2021-11-22 20:42:53,677: WARNING/ForkPoolWorker-1] id=2 | task 1
[2021-11-22 20:42:58,685: WARNING/ForkPoolWorker-1] id=2 | task 2
```

### What I observe: task1 from both the chains are run first, and then the task2 is picked up.

```
[2021-11-22 20:42:43,653: WARNING/ForkPoolWorker-1] id=1 | task 1
[2021-11-22 20:42:48,663: WARNING/ForkPoolWorker-1] id=2 | task 1
[2021-11-22 20:42:53,677: WARNING/ForkPoolWorker-1] id=1 | task 2
[2021-11-22 20:42:58,685: WARNING/ForkPoolWorker-1] id=2 | task 2
```

### I have also tried passing `--concurrency=1` and `--prefetch-multiplier=1` when starting the worker, but it doesn't seem to help, same result as above.

Is this the expected behaviour?

## Setup

- Python: 3.8
- Celery: 5.1.2
- Backend: RabbitMQ
- OS: Linux (Manjaro)

Thanks</div>

**Comment by arshad171** [comment]
any update?

**Comment by glivne** [comment]
You're calling apply_async. why would you assume all tasks of chain 1 would finish before chain 2?

**Comment by arshad171** [comment]
> You're calling apply_async. why would you assume all tasks of chain 1 would finish before chain 2?
> 
> To clarify - if you're looking for these tasks to run all in order, a single chain sounds like a better solution.

can you please suggest the celery API to do that?

my assumption was that `chain(task1, task2)` would link the tasks, and calling `apply_async` would execute that task group asynchronously.
https://docs.celeryproject.org/en/stable/userguide/canvas.html#chains

**Comment by glivne** [comment]
> > You're calling apply_async. why would you assume all tasks of chain 1 would finish before chain 2?
> > To clarify - if you're looking for these tasks to run all in order, a single chain sounds like a better solution.
> 
> can you please suggest the celery API to do that?
> 
> my assumption was that `chain(task1, task2)` would link the tasks, and calling `apply_async` would execute that task group asynchronously. https://docs.celeryproject.org/en/stable/userguide/canvas.html#chains

That is correct. But there is no guarantee between the relative order of tasks outside of each of the chains.
So task1,1 will indeed always execute before task1,2 . And the output tuple args of task 1 will be the first arguments to task 2.
So what you could do, is 1 big chain, if it's okay that task 1,2 will pass its output to task 2,1. Like:

`
chain_result = chain([
 task1.s(1),
 task2.s(),
 task1.s(id=2),
 task2.s()
]).apply_async()
`

If you want to keep the logic separated, you could do a chain of 2 chains. as such:

`
full_chain_result = chain([
 chain(task1.s(1), task2),
 chain(task1.s(2), task2)
]).apply_async()
`
alternatively, you could just wait for chain 1 to finish before calling chain2:
`
res1 = chain(
    task1.s(1),
    task2.s()
).apply_async()
res1.get()
res2 = chain(
    task1.s(2),
    task2.s()
).apply_async()
`

Let me know if any of those worked.

**Comment by arshad171** [comment]
@glivne thank you for the quick response.

The use-case I'm working with involves the client sending independent requests (via REST requests) that need to be processed asynchronously. 
And I have provided a toy example modeling the problem and the result I expect.
- the client sends request to execute a set of tasks (task1 and task2).
- the client expects a async response, so the execution has to be non-blocking.

**solution 1 and 2**: the next request to execute the task chain (task1 and task2) will not be known prior to it's arrival, it can an independent request. hence this will not be possible

solution 3: the tasks I'm performing are very time consuming tasks, and it would better to perform them async. Again the subsequest request to perform the chain (task1, task2) will not be known prior to the client request.

**Comment by arshad171** [comment]
the solution to my problem is simple, "First In First Out" Scheduling, but I'm not sure if celery exposes a way to do this.
